### PR TITLE
feat: form control

### PR DESCRIPTION
**Pull request by voloshynserhii**

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -33,7 +33,8 @@
     "react-dom": "latest",
     "react-idle-timer": "^4.5.0",
     "react-router-dom": "^5.2.0",
-    "react-scripts": "^4.0.1"
+    "react-scripts": "^4.0.1",
+    "validate.js": "^0.13.1"
```
Comment (karpolan) [inline package.json:37]: Nice choice :)

File: src/components/AppInput.js [hunk 2]
```diff
@@ -0,0 +1,38 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+import {FormControl, InputLabel, Input, FormHelperText} from '@material-ui/core';
+import { makeStyles } from '@material-ui/styles';
+
+const useStyles = makeStyles({
+    root: {}
```
Comment (karpolan) [inline src/components/AppInput.js]: Maybe we need a const for that oe remover... not critical

File: src/components/AppInput.js [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+});
+
+const AppInput = ({labelText, inputHelperText, ...props}) => {
+    const classes = useStyles();
+
+    return (
+        <FormControl 
+        className={classes.root}
```
Comment (karpolan) [inline src/components/AppInput.js]: Please format code

use `npm run format'

**Review by karpolan — CHANGES_REQUESTED** [review]
Please form the code

File: src/components/AppInput.js (hunk 3 above)
Comment (voloshynserhii) [reply src/components/AppInput.js]: Done

File: src/components/AppInput.js [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+      disabled={false}
+      fullWidth={true}
+      required={false}
+      size="medium"
+      variant="filled"
+      color="primary"
+      margin="dense"
+    >
+      <InputLabel>{labelText}</InputLabel>
+      <Input name="" {...props} />
```
Comment (karpolan) [inline src/components/AppInput.js:26]: I think passing of `name` property is missing

File: src/components/AppInput.js [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+      <FormHelperText>{inputHelperText}</FormHelperText>
+    </FormControl>
+  );
+};
+
+AppInput.propTypes = {
```
Comment (karpolan) [inline src/components/AppInput.js:32]: I think `name` prop is reqiered because all fields are binded by names

File: src/views/Settings.js [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+  // },
+  // confirmPassword: {
+  //   equality: {
+  //     attribute: 'password',
+  //     // message: 'not match',
+  //   },
+  // }
+};
+
+const Settings = () => {
```
Comment (karpolan) [inline src/views/Settings.js:65]: We use JSDocs for compoents and viesa
/** ...

File: src/views/Settings.js [hunk 7]
```diff
@@ -0,0 +1,184 @@
...
+  // Returns top most Error for the Field by given Name or null
+  const getError = (fieldName) => {
+    return hasError(fieldName) ? formState.errors[fieldName][0] : null;
+  };
+
+  //onChange function
+  const handleChange = useCallback((event) => {
+    event.persist();
+
+    const name = event.target?.name;
```
Comment (karpolan) [inline src/views/Settings.js:93]: A you see name poropert is obligsted

File: src/views/Settings.js [hunk 8]
```diff
@@ -0,0 +1,184 @@
...
+      website: formState.values.website,
+
+      // password: formState.values.password,
+    };
+    console.table(result);
+    return result;
+  };
+
+  return (
+    // <form className={classes.root}>
```
Comment (karpolan) [inline src/views/Settings.js:128]: Remove

File: src/views/Settings.js [hunk 9]
```diff
@@ -0,0 +1,184 @@
...
+      <Grid container xs={12} direction="column" alignItems="center">
+        <Typography className={classes.heading} variant="h2" children="Your Profile" />
+        <Typography className={classes.heading} variant="h6" children="Add information about yourself" />
+      </Grid>
+      <Grid item xs={12} direction="column" justify="space-between" alignItems="center">
+        <AppInput
+          name="nameFirst"
+          value={formState.values.nameFirst || ''}
+          labelText={inputText.name}
+          required={true}
```
Comment (karpolan) [inline src/views/Settings.js:139]: Just `required` is enough for true

File: src/views/Settings.js [hunk 10]
```diff
@@ -0,0 +1,184 @@
...
+  //   equality: {
+  //     attribute: 'password',
+  //     // message: 'not match',
+  //   },
+  // }
+};
+
+const Settings = () => {
+  const classes = useStyles();
+  const [formState, setFormState] = useState(DEFAULT_FORM_STATE);
```
Comment (karpolan) [inline src/views/Settings.js:67]: Adding default values to statre makes innput "managed" by defasult

`useState({...DEFAULT_FORM_STATE, ... {values.name: '', values.email: '...});`

File: src/views/Settings.js [hunk 11]
```diff
@@ -0,0 +1,184 @@
...
+          onChange={handleChange}
+          error={hasError('nameLast')}
+        />
+        <AppInput
+          name="email"
+          value={formState.values.email || ''}
+          labelText={inputText.email}
+          onChange={handleChange}
+          error={hasError('email')}
+          inputHelperText={getError('email')}
```
Comment (karpolan) [inline src/views/Settings.js:158]: `inputHelperText={getError('email') || ' '}`

makes Hlper Text aria always ccupieed

File: src/views/Settings.js [hunk 12]
```diff
@@ -0,0 +1,184 @@
...
+        <AppInput
+          name="website"
+          value={formState.values.website || ''}
+          labelText={inputText.website}
+          inputHelperText={'Print ' + inputText.website.toLowerCase()}
+          onChange={handleChange}
+        />
+      </Grid>
+      <Grid item xs={2}>
+        <AppButton onClick={handleOnConfirm} color="primary" disabled={!formState.isValid}>
```
Comment (karpolan) [inline src/views/Settings.js:176]: order of props should be lphabetical or pops first, events last

File: src/components/AppInput.js [hunk 13]
```diff
@@ -0,0 +1,38 @@
...
+
+const useStyles = makeStyles({
+  root: {},
+});
+
+const AppInput = ({ labelText, inputHelperText, ...props }) => {
+  const classes = useStyles();
+
+  return (
+    <FormControl
```
Comment (karpolan) [inline src/components/AppInput.js:14]: I sugest tot use https://material-ui.com/api/text-field/ as a component base

**Review by karpolan — CHANGES_REQUESTED** [review]

**Pull request closed without merging**